internal/sgbouncewizard/sgbouncewizard.go

```diff
@@ -577,6 +570,50 @@ func (srv *Server) Serve(Port int) {
 		})
 	})
 
+	r.Route("/ws", func(r chi.Router) {
+		r.Get("/", func(w http.ResponseWriter, r *http.Request) {
+			m.HandleRequest(w, r)
+		})
+	})
+
+	m.HandleMessage(func(s *melody.Session, msg []byte) {
+		log.Println(string(msg))
+		params := strings.Split(string(msg), " ")
+		lock.Lock()
+		if params[0] == "edit" {
+			if currentlyViewing[params[1]] {
+				_, exists := s.Get(params[1])
+				if !exists {
+					log.Println("ATTEMPT TO EDIT RULE THAT IS BEING EDITED")
+					s.Write([]byte("INUSE"))
+				} else {
+					s.Write([]byte("ALREADY"))
+				}
+			} else {
+				s.Set(params[1], true)
+				currentlyViewing[params[1]] = true
+				s.Write([]byte("EDIT"))
+				m.BroadcastOthers([]byte("INUSE"), s)
+			}
+		} else if params[0] == "release" {
+			if currentlyViewing[params[1]] {
+				delete(currentlyViewing, params[1])
+				s.Set(params[1], false)
+				m.BroadcastOthers([]byte("FREE"), s)
+			}
+		} else if params[0] == "check" {
+			if currentlyViewing[params[1]] {
+				_, exists := s.Get(params[1])
+				if !exists {
+					s.Write([]byte("INUSE"))
+				}
+			} else {
+				s.Write([]byte("FREE"))
+			}
+		}
+		lock.Unlock()
```

Comment:
`defer lock.Unlock()` at the start of the function call. There is a lot of room for something to panic here (for example, segfaults when `params[1]` doesn't exist) which will cause the handler to terminate immediately and your lock never being freed.